### DISCUSSION TITLE
Modularization Proposal Questions on pull requests

### DISCUSSION THREAD

**Discussion by wipfli** [maplibre/maplibre-native#571] [Q&A]
The modularization proposal #547 by @sjg-wdw has lists 5 pull requests. I copy-paste their descriptions here so people can ask questions on the individual pull requests...

**Comment by wipfli** [comment]
#### PR 1: Shaders

This PR will address shader flexibility.

Sections covered:
* Shader Representation Changes
* Shader Registry

By the end we should have:
* Shaders called out as individual objects accessible outside the toolkit
* Shaders create-able from outside the toolkit
* Shaders replaceable from outside the toolkit
* A compile time option to use shaders either from the compressed source (as now) or with individual shader source files

Tests:
* Create a shader
* Get an existing shader
* Replace an existing shader

**Reply by wipfli** [reply]
I wanted to update the shaders a few months ago to bring them into sync with the latest MapLibre GL JS shaders. It was not easy and did not work out for me. I think there is a lot of custom C++ code for every shader... How will the shader registry take care of this custom C++ stuff?

**Reply by wipfli** [reply]
Like for example here: https://github.com/maplibre/maplibre-gl-native/pull/361

**Reply by sjg-wdw** [reply]
There is some custom logic for each shader.  Mostly it's to do with the attributes and uniforms fed into each.  
In theory we should be able to take the blob, as produced by the tooling and replace it.  I haven't gone through that process, but it's reasonable to add that to one of this PR.

**Reply by wipfli** [reply]
Nice, and can we feed the registry directly with .glsl files like this one: https://github.com/maplibre/maplibre-gl-js/blob/main/src/shaders/circle.fragment.glsl

**Comment by wipfli** [comment]
#### PR 2: Rendering Passes

This PR will add (or restructure) render passes for off and on-screen rendering.

Sections covered:
* Rendering Passes
* Off Screen Render Targets
* Snapshotting

By the end we should have:
* A snapshot callback for one or more frames
* The ability to render specific features to specific render targets
* Off screen render targets that can be read or reused for display

Tests:
* Create an off screen render target
* Direct a set of features to an off screen render target
* Snapshot callback example

**Comment by wipfli** [comment]
#### PR 3: Drawables and Builders

This is the main part of the implementation.  We'll want to shift over to this rendering approach to consolidate the OpenGL portions into a small set of modules that will be easier to duplicate for Metal and other future SDKs.

Sections covered:
* Drawables and Builders

By the end we should have:
* Eliminated the render_layer variants
* Have the layers use Builders to generate Drawables
* Manage and use Drawables in the rendering loop for OpenGL

Tests:
* Output should look the same, or very close.  Use existing render tests.
* Frame rate shouldn't be any lower
* Toolkit shouldn't be any bigger

**Comment by wipfli** [comment]
#### PR 4: Render Loops and Observers

This PR counts as a bit of cleanup, making some of the specific choices we made in the last PR more generic and allowing for outside development of new rendering features.

Sections covered:
* Atlases: Line, Image, Pattern, and Glyph
* Main Render Loops
* Render Observer

By the end we should have:
* A rendering loop that makes very few specific references to outside geometry and texture generators (e.g. the atlases)
* A rendering loop that exposes many of its steps through observer callbacks
* Atlases that may upload their data on other threads

Tests:
* Print observers for all the various callbacks

**Comment by wipfli** [comment]
#### PR 5: Optimization Pass

This PR is here to address areas we may have made slower or bigger in the process of doing our work.  Our goal is to not make things worse, so this is the pass where we try to fix where we did.

Sections covered:
* Optimization

By the end we should have:
* A toolkit that is not any larger, if possible.
* A toolkit that does not start up slower
* A toolkit that renders at least as fast as it used to

Tests:
* Compiled size tests on iOS and Android
* Startup time test on iOS and Android
* Frame rate test on iOS and Android
